sdk/iot/hub/samples/twin/src/iot_hub_twin_example.c

```diff
@@ -0,0 +1,447 @@
+{
+  (void)context;
+
+  if (topicLen == 0)
+  {
+    // The length of the topic if there are one or more NULL characters embedded in topicName,
+    // otherwise topicLen is 0.
+    topicLen = (int)strlen(topicName);
+  }
+
+  az_result result;
+
+  az_span topic_span = az_span_init((uint8_t*)topicName, topicLen);
+
+  az_iot_hub_client_twin_response twin_response;
+  if (az_iot_hub_client_twin_parse_received_topic(&client, topic_span, &twin_response) == AZ_OK)
+  {
+    printf("Twin Message Arrived\n");
+    print_twin_response_type(twin_response.response_type, message);
+    printf("Response status was %i\n", twin_response.status);
+  }
+
+  (void)result;
+
+  putchar('\n');
+  MQTTClient_freeMessage(&message);
+  MQTTClient_free(topicName);
+
+  return 1;
+}
+
+static int connect_device()
+{
+  int rc;
+
+  MQTTClient_SSLOptions mqtt_ssl_options = MQTTClient_SSLOptions_initializer;
+  MQTTClient_connectOptions mqtt_connect_options = MQTTClient_connectOptions_initializer;
+
+  char username[128];
+  size_t username_length;
+  if ((rc = az_iot_hub_client_get_user_name(&client, username, sizeof(username), &username_length))
+      != AZ_OK)
+
+  {
+    printf("Failed to get MQTT clientId, return code %d\n", rc);
+    return rc;
+  }
+
+  mqtt_connect_options.username = username;
+  mqtt_connect_options.password = NULL;
+
+  mqtt_ssl_options.keyStore = (char*)x509_cert_pem_file;
+  if (*x509_trust_pem_file != '\0')
+  {
+    mqtt_ssl_options.trustStore = (char*)x509_trust_pem_file;
+  }
+
+  mqtt_connect_options.ssl = &mqtt_ssl_options;
+
+  if ((rc = MQTTClient_connect(mqtt_client, &mqtt_connect_options)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to connect, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int subscribe()
+{
+  int rc;
+
+  char twin_desired_topic[128];
+  if ((rc = az_iot_hub_client_twin_patch_get_subscribe_topic_filter(
+           &client, twin_desired_topic, sizeof(twin_desired_topic), NULL))
+      != AZ_OK)
+
+  {
+    printf("Failed to get C2D MQTT SUB topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  char twin_response_topic[128];
+  if ((rc = az_iot_hub_client_twin_response_get_subscribe_topic_filter(
+           &client, twin_response_topic, sizeof(twin_response_topic), NULL))
+      != AZ_OK)
+
+  {
+    printf("Failed to get C2D MQTT SUB topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_subscribe(mqtt_client, twin_desired_topic, 1)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to subscribe, return code %d\n", rc);
+    return rc;
+  }
+  if ((rc = MQTTClient_subscribe(mqtt_client, twin_response_topic, 1)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to subscribe, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int send_get_twin()
+{
+  int rc;
+  printf("Requesting twin document\n");
+
+  //Get the topic if we haven't already
+  if (get_twin_topic[0] == '\0')
```

Comment:
This is too complicated for a sample - we should have the sample assume that the topic was not generated.

Reply:
Also, it's likely not OK to submit two requests with the same request_id.

Reply:
From the service docs:

> Request ID can be any valid value for a message property value

So while not entirely useful in this case to have it as the same thing over and over again, it is valid. And since we aren't differentiating between sent/received messages, I think it's okay for now.